2.md

```diff
@@ -0,0 +1,278 @@
+# Virtual Module Sources
+
+## Synopsis
+
+Extend the `Module` constructor such that it accepts virtual module sources:
+objects that implement a protocol that is sufficient for virtualizing the
+evaluation of modules in languages not anticipated by ECMA-262 or a host
+implementations.
+
+## Motivation
+
+Hosts can add support for linking non-EcmaScript languages by creating
+new kinds of ***Module Source Record***.
+For example, a host could add a [[ModuleSource]] internal slot to
+[`WebAssembly.Module`][web-assembly-module] and then link arbitrary WASM into
+an EcmaScript module graph.
+
+However, user code cannot emulate a host in this way.
+To extend or experiment with new languages or new module-like features, user
+code needs a way to emulate or *virtualize* a module source record.
+
+Such a protocol would allow user code to integrate CommonJS, JSON, or WASM
+modules on hosts that do not.
+It would also open a field for experimentation with other kinds of resource
+modules.
+
+## Description
+
+The first argument to the `Module` constructor is a source.
+If that source is an object that does not have a [[Module Source]] internal
+slot, we treat this object as a protocol for loading, binding, linking,
+initializing, and executing the new `Module` instance.
+
+## Interface
+
+```ts
+type VirtualModuleSource = {
+  // Indicates the import and export bindings the module has
+  // between its module environment record, module namespace exotic object,
+  // and its dependencies.
+  bindings?: Array<Binding>,
+
+  // Executes the module if it is imported.
+  // execute may return a promise, indicating that the module uses
+  // the equivalent of top-level-await.
+  execute?: (namespace: ModuleImportsNamespace, {
+    import?: (importSpecifier: string) => Promise<ModuleExportsNamespace>,
+    importMeta?: Object
+  }) => void,
+
+  // Indicates that execute needs to receive a dynamic import function
+  // bound to a Module instance.
+  needsImport?: boolean,
+
+  // Indicates that initialize needs to receive an importMeta.
+  needsImportMeta?: boolean,
+};
+```
+
+Bindings must be one of the shapes proposed in [module source static
+analysis](./1.md), such that for each binding gets linked in a virtual module.
+
+The `Module` constructor from [Module and ModuleSource](./0.md) extends to
+accept virtual module sources instead of `ModuleSource` and reflects whatever
+`source` its given as the `source` property of the returned instance.
+
+## Examples
+
+### JSON
+
+This protocol allows user code to create new module source constructors.
+For example, for a host that does not support JSON, this could be
+accomplished with a small virtual module source.
+
+```js
+class JsonModuleSource {
+  bindings = { export: 'default' };
+  constructor(text) {
+    this.value = JSON.parse(text);
+  };
+  execute(imports) {
+    imports.default = this.value;
+  };
```

Comment:
My understanding is that module sources should be stateless: two `Module` instances with the same module source are equivalent (in a non-virtualized world) to two files with the same contents.
However, this one isn't: if you create two `Module` instances with the same `JsonModuleSource`, you will get the same `.default` object:
```js
const source = new JsonModuleSource(`{ "meaning": 42 }`);

const module1 = new Module(source);
const module2 = new Module(source);

assert(module1 !== module2); // ok

const { default: json1 } = await import(module1);
const { default: json2 } = await import(module1);

assert(json1 !== json2); // fails (they are both the `source.value` object)
```

`JsonModuleSource` should probably be implemented like this:
```js
class JsonModuleSource {
  bindings = { export: 'default' };
  constructor(source) {
    this.source = source;
  };
  execute(imports) {
    imports.default = JSON.parse(this.source);
  };
}
```

I wonder if it's possible to design the API in a way that prevents authors from mistakenly caching the execution result on the module source.

Reply:
It might be possible to discourage but it would not be possible to prevent. And, I’m not sure how to discourage. In this case, I had implemented it as you suggest, but felt like it didn’t capture the sense that the text is not or should not necessarily be retained (as source would be discarded after converted to byte code), and a SyntaxError should occur at compile time and not at runtime.

In this case, I think a better version would parse and then clone.

Reply:
Updated.